yb-voyager/cmd/importSchemaYugabyteDB.go

```diff
@@ -42,44 +42,55 @@ func importSchemaInternal(exportDir string, importObjectList []string,
 }
 
 /*
-Try re-executing each DDL from deffered list.
+Try re-executing each DDL from deferred list.
 If fails, silently avoid the error.
-Else remove from defferedSQLStmts list
+Else remove from deferredSQLStmts list
 At the end, add the unsuccessful ones to a failedSqlStmts list and report to the user
 */
-func importDefferedStatements() {
-	if len(defferedSqlStmts) == 0 {
+func importDeferredStatements() {
+	if len(deferredSqlStmts) == 0 {
 		return
 	}
-	log.Infof("Number of statements in defferedSQLStmts list: %d\n", len(defferedSqlStmts))
+	log.Infof("Number of statements in deferredSQLStmts list: %d\n", len(deferredSqlStmts))
 
 	utils.PrintAndLog("\nExecuting the remaining SQL statements...\n\n")
-	maxIterations := len(defferedSqlStmts)
+	maxIterations := len(deferredSqlStmts)
 	conn := newTargetConn()
 	defer func() { conn.Close(context.Background()) }()
 
 	var err error
 	// max loop iterations to remove all errors
-	for i := 1; i <= maxIterations && len(defferedSqlStmts) > 0; i++ {
-		for j := 0; j < len(defferedSqlStmts); {
-			_, err = conn.Exec(context.Background(), defferedSqlStmts[j].formattedStmt)
+	for i := 1; i <= maxIterations && len(deferredSqlStmts) > 0; i++ {
+		beforeDeferredSqlCount := len(deferredSqlStmts)
+		var failedSqlStmtInIthIteration []string
+		for j := 0; j < len(deferredSqlStmts); j++ {
+			_, err = conn.Exec(context.Background(), deferredSqlStmts[j].formattedStmt)
 			if err == nil {
-				utils.PrintAndLog("%s\n", utils.GetSqlStmtToPrint(defferedSqlStmts[j].stmt))
+				utils.PrintAndLog("%s\n", utils.GetSqlStmtToPrint(deferredSqlStmts[j].stmt))
 				// removing successfully executed SQL
-				defferedSqlStmts = append(defferedSqlStmts[:j], defferedSqlStmts[j+1:]...)
-				break // no increment in j
+				deferredSqlStmts = append(deferredSqlStmts[:j], deferredSqlStmts[j+1:]...)
+				break
 			} else {
-				log.Infof("failed retry of deffered stmt: %s\n%v", utils.GetSqlStmtToPrint(defferedSqlStmts[j].stmt), err)
-				// fails to execute in final attempt
-				if i == maxIterations {
-					errString := "/*\n" + err.Error() + "\n*/\n"
-					failedSqlStmts = append(failedSqlStmts, errString+defferedSqlStmts[j].formattedStmt)
+				log.Infof("failed retry of deferred stmt: %s\n%v", utils.GetSqlStmtToPrint(deferredSqlStmts[j].stmt), err)
+				errString := "/*\n" + err.Error() + "\n*/\n"
+				failedSqlStmtInIthIteration = append(failedSqlStmtInIthIteration, errString+deferredSqlStmts[j].formattedStmt)
+				err = conn.Close(context.Background())
+				if err != nil {
+					log.Warnf("error while closing the connection due to failed deferred stmt: %v", err)
 				}
-				conn.Close(context.Background())
 				conn = newTargetConn()
-				j++
```

Comment:
I'm not sure if it's safe to move the `j++` from here to the for loop definition. Looks like it is here in the else block only because in the event that we successfully execute the stmt, we change the `deferredSqlStmts` array by removing that element, therefore, in the next iteration, the value of `j` should remain unchanged.

Reply:
After this operation of deleting the element from slice: 
```
deferredSqlStmts = append(deferredSqlStmts[:j], deferredSqlStmts[j+1:]...)
```

We also break the inner for-loop, so there is no next iteration.

Let me know if i am missing anything.

Reply:
Oh right, I see the break! 
But now I'm confused; why do we break when we successfully execute a deferred statement? Shouldn't we continue?

Reply:
once the jth statement executed successfully, there are chances that some of the ddls before 'j' were dependent on 'j' and can be executed successfully, so as soon as 'j' passes we start over.